var/spack/repos/builtin/packages/spark/package.py

```diff
@@ -68,6 +68,7 @@ def install_dir(dirname):
     @when('+hadoop')
     def setup_environment(self, spack_env, run_env):
         hadoop = self.spec['hadoop'].command
+        hadoop.add_default_env('JAVA_HOME', self.spec['java'].home)
```

Comment:
shouldn't this be set by the Hadoop package?

Reply:
JAVA_HOME is set by jdk's setup_dependent_environment and hadoop has java as a runtime dependency. But I didn't think the spack.spec.command SpecBuildInterface passed the package's runtime environment into the executable that is returned.

Though I do think it makes more sense for spec['hadoop'].command to return an Executable instance with the runtime environment set already. 

The `spec._default_command_handler` just returns Executable(prefix_path). If I'm reading this right, however, we can just add a 

```
@property
def command(self):
    bin = Executable(self.prefix)
    bin.add_default_env('JAVA_HOME', self.spec['java'].home)
    return bin
```

to the hadoop package? Or should the SpecBuildInterface update the Executable's env with the packages's runtime environment by default?

Reply:
@mpbelhorn your suggestion would work. But I think the way you do things here currently is good.

In most cases Spack and the Spack Java packages will set up JAVA_HOME appropriately for all commands that are executed as part of the build. In this case you are executing hadoop as part of the setup function which is a bit different.

I'm not sure what more `SPARK_DIST_CLASSPATH` would need beyond what the Spack Java package already inserts into `CLASSPATH`. So ideally it would either not be required or the CLASSPATH logic in Spack's Java packages would be refactored into a function that other Java deps (like `Spark`) could call.